### ISSUE TITLE
Epic: TapDance and its many flavors

### ISSUE THREAD

**Issue by kdb424** [KMKfw/kmk_firmware#40]
Add support for tap dance features allowing users to define custom actions based on different counts of taps or tap and hold.

This will be achieved by setting a timer in the state (user configurable, possibly per key), and when the timer expires, it will act as a macro, that fires other macros.

- TD(KC.SPACE, KC.LEAD)
- TD(KC.U, send_string('test'))
- TD(MO(2), MO(3))

Those should all be valid. Depending on what is received, the tap dance will handle differently. For example, things like MO(2) would be fired on keydown as there is no output to suppress and offers the best experience having instant results where possible. If something like KC.SPACE or any other keycode that would trigger HID, or a HID changing action, such as KC.LEAD, would be fired ONLY after the timer expires.

Internal keycodes should have a tag on them that would offer unblocking of tap dance, or it will wait until  final resolution of the timer to facilitate easy additions of keycodes, and ensure that tap dance will function properly with all new additions by default.

With this addition, keycodes such as TT could simply be written by using 
TD(MO(2), TG(2))

**Comment by kdb424** [comment]
I will consider this "good enough" at this point on the mainline branch that this is no longer needed.
